Cesium3DTilesSelection/src/Tile.cpp

```diff
@@ -297,21 +293,46 @@ void Tile::loadContent() {
                   pRequest->url());
               auto pLoadResult = std::make_unique<TileContentLoadResult>();
               pLoadResult->httpStatusCode = pResponse->statusCode();
-              return LoadResult{
-                  LoadState::FailedTemporarily,
-                  std::move(pLoadResult),
-                  nullptr};
+              return asyncSystem.createResolvedFuture(std::move(pLoadResult));
             }
 
-            loadInput.data = pResponse->data();
-            loadInput.contentType = pResponse->contentType();
-            loadInput.url = pRequest->url();
-            std::unique_ptr<TileContentLoadResult> pContent =
-                TileContentFactory::createContent(loadInput);
-
+            loadInput.asyncSystem = asyncSystem;
+            loadInput.pLogger = std::move(pLogger);
+            loadInput.pAssetAccessor = std::move(pAssetAccessor);
+            loadInput.pRequest = std::move(pRequest);
+
+            return TileContentFactory::createContent(loadInput)
+                // Forward status code to the load result.
+                .thenImmediately(
+                    [statusCode = pResponse->statusCode()](
+                        std::unique_ptr<TileContentLoadResult>&& pContent) {
+                      if (pContent) {
+                        pContent->httpStatusCode = statusCode;
+                      }
+
+                      return std::move(pContent);
+                    });
+          })
+      .thenInWorkerThread(
+          [gltfUpAxis,
+           &boundingVolume = this->_boundingVolume,
+           &transform = this->_transform,
```

Comment:
These references are a bit unsafe. I mean, it's probably mostly fine as the code is currently written, but it's kind of a policy violation that might lead to (subtle, hard to debug) bugs in the future.

The violated policy is: never share not-thread-safe objects between threads. The bounding volume and transform are members of the Tile, and Tile is not thread safe. So sharing a reference to them with the worker thread means multiple threads can end up accessing them simultaneously. Now that's technically fine as long as neither thread is _modifying_ either field, and I think the code as currently written will not modify a Tile's transform or bounding volume while that Tile is in the process of being loaded. So it's not an immediate problem.

But this is fragile, because it means in all our modifications in the future, we have to continue to make sure that a Tile's bounding volume and transform are not modified while an async Load is in progress. And if we fail to do so, it'll probably mostly continue to work just fine, unless we just happen to have some really bad timing and then things will start failing in super mysterious ways (e.g. totally wrong texture coordinates in rare circumstances because we used a half-updated transform to generate them).

We have a policy against _ever_ sharing not-thread-safe objects between threads in order to avoid the need to ever think "will this be modified?" That's a hard enough question to answer for the code base today, and it's even harder to ensure our answer doesn't change in the future.

The only exception should be when the cost of copying is too high, and we can't see another way to go about it. I don't think we currently have any cases like that.

I don't think a bounding volume or transform are so expensive that we need to stress too much about copying them. One way we could avoid it, though, (which may be worth considering), is by making the second `thenInWorkerThread` here inside of the other one, as a continuation of the `thenImmediately` above. That would allow the `TileContentLoadInput` to be moved in, making the code overall much more similar to how it was in main. But at the cost of some extra indentation.

Reply:
@kring I combined the second `thenInWorkerThread` with the nested `thenImmediately` and it looks cleaner now, is this what you were thinking? It looks like `loadInput` still needs to be captured-by-reference to avoid a copy, but I think this is much safer than capturing the Tile members directly by reference. The `loadInput` can't be moved into the lambda capture since it will still be used by the `createContent` invocation, but by the time the `thenInWorkerThread` runs, `createContent` is clearly done, so the `loadInput` reference will be safe to use. Mostly just expounding on this to make sure it doesn't violate some policy like before.

Reply:
Capturing the loadInput by reference is not safe. That object is owned by the outer lambda, so you're relying on the lifetime of that outer lambda extending until after the inner lambda completes. I'm actually somewhat surprised that's working. I thought Async++ destroys the function immediately after it returns. Even if it doesn't, that's an implementation detail we shouldn't rely on.

I should be safe and reasonable to move loadInput into the inner lambda. Yes, it also gets passed to `TileContentFactory::createContent` by reference, but that happens first, so it should be fine to move the instance into the lambda afterward. `TileContentFactory::createContent` isn't allowed to hold on to and use that reference past the point that it returns, consistent with normal rules for reference parameters, so that shouldn't stop it. A `createContent` lambda capturing that loadInput parameter by reference would be unsafe in the same way that it's unsafe for the lsambda here to do so. If `createContent` needs some or all of the TileLoadInput later, it needs to make a copy.